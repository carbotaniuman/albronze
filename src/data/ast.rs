#[derive(Clone, Debug, PartialEq)]
pub enum ExternalDeclaration {
    Function,
    Declaration,
}
