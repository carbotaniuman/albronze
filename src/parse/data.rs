use crate::InternedStr;

#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum Keyword {
    // statements
    If,
    Else,
    Do,
    While,
    For,
    Switch,
    Case,
    Default,
    Break,
    Continue,
    Return,
    Goto,

    // types
    Char,
    Short,
    Int,
    Long,
    Float,
    Double,
    Void,
    Signed,
    Unsigned,
    Typedef,

    // user-defined types
    Union,
    Struct,
    Enum,
    // the `i` in `typedef int i;`
    UserTypedef(InternedStr),

    // weird types
    Bool,
    Complex,
    Imaginary,
    VaList,

    // qualifiers
    Const,
    Volatile,
    Restrict,
    // weird qualifiers
    Atomic,
    ThreadLocal,
    // function qualifiers
    Inline,
    NoReturn,

    // storage classes
    Auto,
    Register,
    Static,
    Extern,

    // intrinsics
    Sizeof,
    Generic,
    StaticAssert,
    Alignas,
    Alignof,
}

impl std::fmt::Display for Keyword {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            Keyword::Alignas
            | Keyword::Alignof
            | Keyword::Bool
            | Keyword::Complex
            | Keyword::Imaginary
            | Keyword::Atomic
            | Keyword::Generic => write!(f, "_{:?}", self),
            Keyword::NoReturn => write!(f, "_Noreturn"),
            Keyword::ThreadLocal => write!(f, "_Thread_local"),
            Keyword::StaticAssert => write!(f, "_Static_assert"),
            Keyword::VaList => write!(f, "va_list"),
            _ => write!(f, "{}", &format!("{:?}", self).to_lowercase()),
        }
    }
}
