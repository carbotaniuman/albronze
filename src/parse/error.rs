use crate::preprocess::TokenKind;

// use super::data::Keyword;

use thiserror::Error;

/// Syntax errors are non-exhaustive and may have new variants added at any time
#[derive(Clone, Debug, Error, PartialEq)]
#[non_exhaustive]
pub enum SyntaxError {
    // TODO: replace this
    #[error("{0}")]
    Generic(String),

    #[error("expected {0}, got <end-of-file>")]
    EndOfFile(&'static str),

    #[error("expected statement, got {0}")]
    NotAStatement(super::Keyword),

    // expected a primary expression, but got EOF or an invalid token
    #[error("expected variable, literal, or '('")]
    MissingPrimary,

    #[error("expected identifier, got '{}'",
        .0.as_ref().map_or("<end-of-file>".into(),
                           |t| std::borrow::Cow::Owned(t.to_string())))]
    ExpectedId(Option<TokenKind>),

    #[error("expected declaration specifier, got keyword '{0}'")]
    ExpectedDeclSpecifier(super::Keyword),

    #[error("expected declarator in declaration")]
    ExpectedDeclarator,

    #[error("empty type name")]
    ExpectedType,

    #[error("expected '(', '*', or variable, got '{0}'")]
    ExpectedDeclaratorStart(TokenKind),

    #[error("only functions can have a function body (got {0})")]
    NotAFunction(super::ast::InitDeclarator),

    #[error("functions cannot be initialized (got {0})")]
    FunctionInitializer(super::ast::Initializer),

    #[error("function not allowed in this context (got {})", .0.as_type())]
    FunctionNotAllowed(super::ast::FunctionDefinition),

    #[error("function definitions must have a name")]
    MissingFunctionName,

    #[error("`static` for array sizes is only allowed in function declarations")]
    StaticInConcreteArray,
}
