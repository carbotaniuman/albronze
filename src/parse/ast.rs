use crate::data::{AssignmentToken, ComparisonToken, LiteralValue};
use crate::location::Locatable;
use crate::InternedStr;

use std::fmt;

use derive_more::From;

#[derive(Clone, Debug, PartialEq)]
pub enum ExternalDeclaration {
    Function(FunctionDefinition),
    Declaration(Declaration),
}

#[derive(Clone, Debug, PartialEq)]
pub struct FunctionDefinition {
    pub specifiers: Vec<DeclarationSpecifier>,
    pub id: InternedStr,
    pub declarator: FunctionDeclarator,
    pub body: CompoundStatement,
}

impl FunctionDefinition {
    pub(crate) fn as_type(&self) -> TypeName {
        TypeName {
            specifiers: self.specifiers.clone(),
            declarator: Declarator {
                decl: DeclaratorType::Function(self.declarator.clone()),
                id: Some(self.id),
            },
        }
    }
}

pub type Program = Vec<Declaration>;

#[derive(Clone, Debug, PartialEq)]
pub struct Declaration {
    pub specifiers: Vec<DeclarationSpecifier>,
    pub declarators: Vec<Locatable<InitDeclarator>>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct TypeName {
    pub specifiers: Vec<DeclarationSpecifier>,
    pub declarator: Declarator,
}

impl fmt::Display for TypeName {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        todo!()
    }
}

#[derive(Clone, Debug, PartialEq, From)]
pub enum DeclarationSpecifier {
    #[from]
    Unit(UnitSpecifier),
    Struct(StructSpecifier),
    Union(StructSpecifier),
    // enum name? { A = 1, B = 2, C }
    Enum {
        name: Option<InternedStr>,
        members: Option<Vec<(InternedStr, Option<Expr>)>>,
    },
    // NOTE: _not_ the same as UnitSpecifier::Typedef
    // that represents the `typedef` keyword, this represents a name that has been typedef-ed
    Typedef(InternedStr),
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
pub enum UnitSpecifier {
    // types
    Char,
    Short,
    Int,
    Long,
    Float,
    Double,
    Void,
    Signed,
    Unsigned,

    // weird types
    Bool,
    Complex,
    Imaginary,
    VaList,

    // qualifiers
    Const,
    Volatile,
    Restrict,
    // weird qualifiers
    Atomic,
    ThreadLocal,
    // function qualifiers
    Inline,
    NoReturn,

    // storage classes
    Auto,
    Register,
    Static,
    Extern,
    Typedef,
}

#[derive(Clone, Debug, PartialEq)]
pub struct StructSpecifier {
    pub name: Option<InternedStr>,
    /// Some([]): `struct s {}`
    /// None: `struct s;`
    pub members: Option<Vec<StructDeclarationList>>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct StructDeclarationList {
    pub specifiers: Vec<DeclarationSpecifier>,
    pub declarators: Vec<StructDeclarator>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct StructDeclarator {
    /// optional since this could be only padding bits
    pub decl: Option<Declarator>,
    pub bitfield: Option<Expr>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct InitDeclarator {
    pub init: Option<Initializer>,
    pub declarator: Declarator,
}

impl fmt::Display for InitDeclarator {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        todo!()
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum Initializer {
    Scalar(Box<Expr>),
    Aggregate(Vec<Initializer>),
}

impl fmt::Display for Initializer {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        todo!()
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct Declarator {
    pub decl: DeclaratorType,
    pub id: Option<InternedStr>,
}

impl Declarator {
    fn is_nonempty(&self) -> bool {
        match self.decl {
            DeclaratorType::End => self.id.is_some(),
            _ => true,
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct FunctionDeclarator {
    pub return_type: Box<DeclaratorType>,
    // TODO: maybe support K&R C?
    //DeclarationList
    pub params: Vec<TypeName>,
    pub varargs: bool,
}

#[derive(Clone, Debug, PartialEq, From)]
pub enum DeclaratorType {
    // No more declarator, e.g. for abstract params
    End,
    Pointer {
        to: Box<DeclaratorType>,
        qualifiers: Vec<DeclarationSpecifier>,
    },
    Array {
        of: Box<DeclaratorType>,
        size: Option<Box<Expr>>,
    },
    #[from]
    Function(FunctionDeclarator),
}

pub type Stmt = Locatable<StmtType>;
pub type CompoundStatement = Vec<Stmt>;

#[derive(Clone, Debug, PartialEq)]
pub enum StmtType {
    Compound(CompoundStatement),
    If(Expr, Box<Stmt>, Option<Box<Stmt>>),
    Do(Box<Stmt>, Expr),
    While(Expr, Box<Stmt>),
    // for(int i = 1, j = 2; i < 4; ++i) body
    // for(i = 1; ; ++i) body
    // for (;;) ;
    For {
        initializer: Box<Stmt>,
        condition: Option<Box<Expr>>,
        post_loop: Option<Box<Expr>>,
        body: Box<Stmt>,
    },
    Switch(Expr, Box<Stmt>),
    Label(InternedStr, Box<Stmt>),
    Case(Box<Expr>, Box<Stmt>),
    Default(Box<Stmt>),
    Expr(Expr),
    Goto(InternedStr),
    Continue,
    Break,
    Return(Option<Expr>),
    Decl(Declaration),
}

pub type Expr = Locatable<ExprType>;

#[derive(Clone, Debug, PartialEq)]
pub enum ExprType {
    // primary
    Id(InternedStr),
    Literal(LiteralValue),

    // postfix
    FuncCall(Box<Expr>, Vec<Expr>),
    Member(Box<Expr>, InternedStr),
    DerefMember(Box<Expr>, InternedStr),
    // post increment/decrement
    PostIncrement(Box<Expr>, bool),
    // a[i]
    Index(Box<Expr>, Box<Expr>),

    // prefix
    PreIncrement(Box<Expr>, bool),
    Cast(TypeName, Box<Expr>),
    AlignofType(TypeName),
    AlignofExpr(Box<Expr>),
    SizeofType(TypeName),
    SizeofExpr(Box<Expr>),
    Deref(Box<Expr>),
    AddressOf(Box<Expr>),
    UnaryPlus(Box<Expr>),
    Negate(Box<Expr>),
    BitwiseNot(Box<Expr>),
    LogicalNot(Box<Expr>),

    // binary
    LogicalOr(Box<Expr>, Box<Expr>),
    BitwiseOr(Box<Expr>, Box<Expr>),
    LogicalAnd(Box<Expr>, Box<Expr>),
    BitwiseAnd(Box<Expr>, Box<Expr>),
    Xor(Box<Expr>, Box<Expr>),
    Mul(Box<Expr>, Box<Expr>),
    Div(Box<Expr>, Box<Expr>),
    Mod(Box<Expr>, Box<Expr>),
    Add(Box<Expr>, Box<Expr>),
    Sub(Box<Expr>, Box<Expr>),
    // bool: left or right
    Shift(Box<Expr>, Box<Expr>, bool),
    // Token: make >, <, <=, ... part of the same variant
    Compare(Box<Expr>, Box<Expr>, ComparisonToken),
    // Token: allow extended assignment
    Assign(Box<Expr>, Box<Expr>, AssignmentToken),

    // misfits
    // Ternary: if ? then : else
    Ternary(Box<Expr>, Box<Expr>, Box<Expr>),
    Comma(Box<Expr>, Box<Expr>),
}

impl Default for StmtType {
    fn default() -> Self {
        StmtType::Compound(Vec::new())
    }
}

