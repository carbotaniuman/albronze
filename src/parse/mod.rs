mod ast;
mod decl;
mod error;
mod expr;
mod stmt;

use crate::error::{ErrorHandler, Warning};
use crate::location::{Locatable, Location};
use crate::parse::error::SyntaxError;
use crate::preprocess::{EncodingKind, Keyword, LexResult, LiteralKind, TokenKind};
use crate::scope::Scope;
use crate::InternedStr;
use ast::ExternalDeclaration;

use std::collections::VecDeque;
use std::mem;

type Lexeme = LexResult<Locatable<TokenKind>>;
type SyntaxResult<T> = Result<T, Locatable<SyntaxError>>;

pub struct Parser<I: Iterator<Item = Lexeme>> {
    /// Parser hack so that we know that `typedef int i; i j;`
    /// is legal, and also to resolve ambiguities
    typedefs: Scope<InternedStr, ()>,
    /// VecDeque supports pop_front with reasonable efficiency
    /// this is useful because there could be multiple declarators
    /// in a single declaration; e.g. `int a, b, c;`
    pending: VecDeque<Locatable<ExternalDeclaration>>,

    /// Whether to debug each declaration
    debug: bool,
    /// in case we get to the end of the file and want to show an error
    last_location: Location,
    /// Internal API which makes it easier to return errors lazily
    error_handler: ErrorHandler<SyntaxError>,

    /// We iterate lazily over the tokens, so if we have a program that's mostly valid but
    /// breaks at the end, we don't only show lex errors
    tokens: std::iter::Peekable<I>,
    /// the last token we saw from the Lexer. None if we haven't looked ahead.
    /// Should only be used in this module.
    current: Option<Locatable<TokenKind>>,
    /// TODO: are we sure we need 2 tokens of lookahead?
    /// this was put here for declarations, so we know the difference between
    /// int (*x) and int (int), but there's probably a workaround
    next: Option<Locatable<TokenKind>>,
}

impl<I: Iterator<Item = Lexeme>> Parser<I> {
    pub fn new(tokens: I, debug: bool) -> Self {
        Parser {
            typedefs: Default::default(),
            tokens: tokens.peekable(),
            pending: Default::default(),
            // The only time this is used is when an error occurs,
            // which only happens after at least one token has been seen.
            // So this default location will never be used.
            last_location: Location::default(),
            current: None,
            next: None,
            debug,
            error_handler: ErrorHandler::new(),
            // recursion_guard: Default::default(),
        }
    }

    /// Return whether this parser has fully finished parsing.
    ///
    /// This can be used if, for example, you call `parser.expr()`
    /// and want to see if there are any left-over tokens.
    pub fn is_empty(&mut self) -> bool {
        self.peek_token().is_none()
    }
}

impl<I: Iterator<Item = Lexeme>> Iterator for Parser<I> {
    type Item = SyntaxResult<Locatable<ExternalDeclaration>>;
    /// ```yacc
    /// translation_unit
    /// : external_declaration
    /// | translation_unit external_declaration
    /// ;
    ///
    /// external_declaration
    /// : function_definition
    /// | declaration
    /// ;
    ///
    /// function_definition
    /// : declarator compound_statement
    /// | declaration_specifiers declarator compound_statement
    /// ;
    /// ```
    /// <http://www.quut.com/c/ANSI-C-grammar-y.html#translation_unit>
    fn next(&mut self) -> Option<Self::Item> {
        loop {
            // check for pending changes from the last declaration
            if let Some(err) = self.error_handler.pop_error() {
                return Some(Err(err));
            } else if let Some(decl) = self.pending.pop_front() {
                if self.debug {
                    println!("ast: {:?}", decl.data);
                }
                return Some(Ok(decl));
            }

            // Check for end of file
            if self.peek_token().is_none() {
                // peek_token can encounter lex errors that it doesn't return
                return self.error_handler.pop_error().map(Err);
            }

            // Remove extra semicolons
            while let Some(locatable) = self.match_next(&TokenKind::Semicolon) {
                self.error_handler.warn(
                    Warning::ExtraneousSemicolon("top level"),
                    locatable.location,
                );
            }

            // Parse more of our file
            match self.external_declaration() {
                Ok(decls) => {
                    self.pending.push_back(decls);
                }
                Err(err) => {
                    // there could be semantic errors that were reported in the meantime,
                    // so we can't just return this error (it might be in the wrong order)
                    self.error_handler.push_error(err);
                }
            }
        }
    }
}

impl<I: Iterator<Item = Lexeme>> Parser<I> {
    // fn recursion_check(&mut self) -> RecursionGuard {
    //     self.recursion_guard
    //         .recursion_check(&mut self.error_handler)
    // }
    // don't use this, use next_token instead
    // WARNING: this _cannot_ read or modify `self.current` or `self.next`
    fn __impl_next_token(&mut self) -> Option<Locatable<TokenKind>> {
        loop {
            match self.tokens.next() {
                Some(Ok(Locatable {
                    data: TokenKind::Whitespace(_),
                    ..
                })) => continue,
                // Some(Ok(Locatable {
                //     data: TokenKind::Literal(LiteralToken::Str(mut concat_strs)),
                //     mut location,
                // })) => {
                //     // 5.1.1.2p1: Translation phase 6: Adjacent string literal tokens are concatenated.
                //     loop {
                //         match self.tokens.peek() {
                //             Some(Ok(Locatable {
                //                 data: TokenKind::Literal(LiteralToken::Str(merge_strs)),
                //                 location: next_location,
                //             })) => {
                //                 location = location.merge(next_location);
                //                 concat_strs.append(&mut merge_strs.clone());
                //                 self.tokens.next(); // Actually remove next
                //             }
                //             Some(Ok(Locatable {
                //                 data: TokenKind::Whitespace(_),
                //                 ..
                //             })) => {
                //                 self.tokens.next();
                //             }
                //             Some(Ok(_)) => break,
                //             Some(Err(_)) => {
                //                 let err = self.tokens.next().unwrap().unwrap_err();
                //                 self.error_handler.push_back(err);
                //             }
                //             None => break,
                //         }
                //     }
                //     break Some(Locatable::new(
                //         TokenKind::Literal(LiteralToken::Str(concat_strs)),
                //         location,
                //     ));
                // }
                Some(Ok(token)) => {
                    self.last_location = token.location;
                    // This is _such_ a hack
                    // I'd much rather use `Token::is_decl_specifier()` at the various places it's necessary,
                    // but that runs into limits of the lifetime system since `peek_token()` takes `&mut self`:
                    // https://doc.rust-lang.org/nomicon/lifetime-mismatch.html#limits-of-lifetimes
                    if let TokenKind::Identifier(id) = token.data {
                        if self.typedefs.get(&id).is_some() {
                            break Some(
                                token
                                    .location
                                    .with(TokenKind::Keyword(Keyword::UserTypedef(id))),
                            );
                        }
                    }
                    break Some(token);
                }
                // Some(Err(err)) => {
                //     self.last_location = err.location();
                //     self.lex_error(err);
                // }
                None => break None,
                _ => todo!(),
            }
        }
    }
    fn next_token(&mut self) -> Option<Locatable<TokenKind>> {
        mem::replace(&mut self.current, self.next.take()).or_else(|| self.__impl_next_token())
    }
    fn peek_token(&mut self) -> Option<&TokenKind> {
        if self.current.is_none() {
            self.current = self.next.take().or_else(|| self.__impl_next_token());
        }
        self.current.as_ref().map(|x| &x.data)
    }
    // TODO: this is mostly copied from peek_token
    fn peek_next_token(&mut self) -> Option<&TokenKind> {
        if self.next.is_none() {
            if self.current.is_none() {
                self.current = self.__impl_next_token();
            }
            self.next = self.__impl_next_token();
        }
        self.next.as_ref().map(|x| &x.data)
    }
    fn next_location(&self) -> Location {
        if let Some(token) = &self.current {
            token.location
        } else {
            self.last_location
        }
    }
    fn match_id(&mut self) -> Option<Locatable<InternedStr>> {
        match self.peek_token() {
            Some(&TokenKind::Identifier(name))
            | Some(&TokenKind::Keyword(Keyword::UserTypedef(name))) => {
                let location = self.next_token().unwrap().location;
                Some(Locatable::new(name, location))
            }
            _ => None,
        }
    }
    fn match_keywords(&mut self, keywords: &[Keyword]) -> Option<Locatable<Keyword>> {
        if let Some(&TokenKind::Keyword(keyword)) = self.peek_token() {
            for expected in keywords {
                if keyword == *expected {
                    let location = self.next_token().unwrap().location;
                    return Some(Locatable::new(keyword, location));
                }
            }
            None
        } else {
            None
        }
    }
    fn match_literal(&mut self) -> Option<Locatable<(LiteralKind, InternedStr)>> {
        let next = self.next_token();
        if let Some(Locatable {
            data: TokenKind::Literal(lit, interned),
            location,
        }) = next
        {
            Some(location.with((lit, interned)))
        } else {
            self.unput(next);
            None
        }
    }
    fn match_string_literal(&mut self) -> Option<Locatable<(EncodingKind, InternedStr)>> {
        let next = self.next_token();
        if let Some(Locatable {
            data: TokenKind::Literal(LiteralKind::String(e), interned),
            location,
        }) = next
        {
            Some(location.with((e, interned)))
        } else {
            self.unput(next);
            None
        }
    }
    fn match_next(&mut self, next: &TokenKind) -> Option<Locatable<TokenKind>> {
        self.match_any(&[next])
    }
    fn match_any(&mut self, choices: &[&TokenKind]) -> Option<Locatable<TokenKind>> {
        if let Some(data) = self.peek_token() {
            for token in choices {
                if token.same_kind(data) {
                    return self.next_token();
                }
            }
        }
        None
    }
    /*
     * If we're in an invalid state, try to recover.
     * Consume tokens until the end of a statement - either ';' or '}'
     */
    fn panic(&mut self) {
        while let Some(token) = self.next_token() {
            match token.data {
                TokenKind::Semicolon => break,
                TokenKind::RightBrace(_) => {
                    break;
                }
                _ => continue,
            };
        }
    }
    fn expect_id(&mut self) -> SyntaxResult<Locatable<InternedStr>> {
        if let Some(id) = self.match_id() {
            Ok(id)
        } else {
            let err = Locatable {
                data: SyntaxError::ExpectedId(self.peek_token().cloned()),
                location: self.next_location(),
            };
            self.panic();
            Err(err)
        }
    }
    fn expect(&mut self, next: TokenKind) -> SyntaxResult<Locatable<TokenKind>> {
        let token = match self.peek_token() {
            Some(t) => t,
            None => {
                let err = Locatable {
                    data: SyntaxError::Generic(format!("expected '{}', got '<end-of-file>'", next)),
                    // TODO: we don't actually want this, we want the end of the file
                    location: self.last_location,
                };
                self.panic();
                return Err(err);
            }
        };
        if token.same_kind(&next) {
            Ok(self.next_token().unwrap())
        } else {
            let err = Locatable {
                data: SyntaxError::Generic(format!("expected '{}', got '{}'", next, token)),
                location: self.next_location(),
            };
            self.panic();
            Err(err)
        }
    }
    /// - replace `self.current` with `item`
    /// - replace `self.next` with `self.current`
    /// - the previous value of `self.next` is lost
    fn unput(&mut self, item: Option<Locatable<TokenKind>>) {
        assert!(self.next.is_none());
        self.next = mem::replace(&mut self.current, item);
    }
    fn lex_error(&mut self, err: Locatable<SyntaxError>) {
        self.error_handler.push_error(err);
    }
    pub fn collect_results(
        &mut self,
    ) -> (
        Vec<Locatable<ExternalDeclaration>>,
        Vec<Locatable<SyntaxError>>,
    ) {
        let mut decls = Vec::new();
        let mut errs = Vec::new();
        for result in self {
            match result {
                Ok(decl) => decls.push(decl),
                Err(err) => errs.push(err),
            }
        }
        (decls, errs)
    }
    /// Return all warnings seen so far.
    ///
    /// These warnings are consumed and will not be returned if you call
    /// `warnings()` again.
    pub fn warnings(&mut self) -> VecDeque<Locatable<Warning>> {
        self.error_handler.take_warnings()
    }
}
