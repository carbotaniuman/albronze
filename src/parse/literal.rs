use super::*;
use crate::parse::ast::{FloatRadix, LiteralData, IntegerRadix};

use thiserror::Error;

#[derive(Clone, Debug, Error, PartialEq)]
pub enum NumericLiteralError {
    #[error("hexadecimal floating point literals require an exponent")]
    MissingHexExponent,

    #[error("exponent is missing digits")]
    EmptyExponent,
}

pub fn number_literal(value: &str) -> Result<LiteralData, NumericLiteralError> {
    // So, integer literals in C are actually pretty complicated, given
    // hexadecimal, integral, and octal. We do parsing in a linear
    // scan fashion here.

    let is_hexdigit = |f: char| f.is_ascii_hexdigit() || f == '.';
    let is_digit = |f: char| f.is_ascii_digit() || f == '.';

    #[derive(Copy, Clone, PartialEq, Eq, Debug)]
    enum Mode {
        Hex,
        Oct,
        Dec,
    }

    impl Mode {
        fn chars_to_skip(self) -> usize {
            match self {
                Mode::Hex => 2,
                Mode::Oct => 1,
                Mode::Dec => 0,
            }
        }

        fn to_integer_radix(self) -> IntegerRadix {
            match self {
                Mode::Hex => IntegerRadix::Hexadecimal,
                Mode::Oct => IntegerRadix::Octal,
                Mode::Dec => IntegerRadix::Decimal,
            }
        }

        fn to_float_radix(self) -> FloatRadix {
            match self {
                Mode::Hex => FloatRadix::Hexadecimal,
                Mode::Dec => FloatRadix::Decimal,
                Mode::Oct => unreachable!(),
            }
        }
    }

    // Figure out what kind of 
    let mut mode = if value.starts_with("0x") || value.starts_with("0X") {
        Mode::Hex
    } else if value.starts_with("0") {
        Mode::Oct
    } else {
        Mode::Dec
    };

    // This (along with `mode` above), may get changed depending on
    // what the rest of the scan reveals. For instance, a lone `0` is `0`,
    // not an empty octal constant, while `0.7` or `0.` are float literals. 
    let mut rest = &value[mode.chars_to_skip()..];

    #[derive(Copy, Clone, PartialEq, Eq, Debug)]
    enum State {
        Integer,
        Fraction,
        Exponent,
        Suffix,
    }

    // finds the end of the integers, but is `None` if `rest` has only integers
    // remaining, so it's really the start of the fraction, exponent, or suffix.
    let non_integer_start = match mode {
        Mode::Hex => rest.find(|x: char| !x.is_ascii_hexdigit()),
        Mode::Oct if rest.is_empty() => {
            // if `rest` is empty, then we parsed a 0 as an octal,
            // instead of a plain 0. fix that here
            mode = Mode::Dec;
            rest = value;
            None
        },
        Mode::Oct => rest.find(|x: char| !x.is_ascii_digit()),
        Mode::Dec => rest.find(|x: char| !x.is_ascii_digit()),
    };

    // after this, we know that numbers must have a fraction,
    // exponent, or suffix
    let Some(mut non_integer_start) = non_integer_start else {
        return Ok(LiteralData::Integer {
            value: rest.into(),
            suffix: InternedStr::default(),
            radix: mode.to_integer_radix(),
        });
    };

    // from here on out, we'll just reuse state and rest,
    // and reassign them to deal with all the possible conditions
    // that may occur.
    let (is_float, integer, mut rest, mut state) = 'n: {
                let remainder = &rest[non_integer_start..];

        let new_state = if remainder.starts_with(".") {
            State::Fraction
        } else if remainder.starts_with("p") || remainder.starts_with("P") {
            debug_assert_eq!(mode, Mode::Hex);
            State::Exponent
        } else if remainder.starts_with("e") || remainder.starts_with("E") {
            debug_assert_ne!(mode, Mode::Hex);
            State::Exponent
        } else {
            // We are not a float, we don't need to do any fixups
            break 'n (false, &rest[0..non_integer_start], remainder, State::Suffix)
        };

        if mode == Mode::Oct {
            mode = Mode::Dec;
            rest = value;
            non_integer_start += 1;
        };
        (true, &rest[0..non_integer_start], remainder, new_state)
    };

    let fraction = if state == State::Fraction {
        let fraction_part = &rest[1..];
        let non_fraction_start = match mode {
            Mode::Hex => fraction_part.find(|x: char| !x.is_ascii_hexdigit()),
            Mode::Dec => fraction_part.find(|x: char| !x.is_ascii_digit()),
            Mode::Oct => unreachable!(),
        };

        let Some(non_fraction_start) = non_fraction_start else {
            if mode == Mode::Hex {
                return Err(NumericLiteralError::MissingHexExponent);
            }

            return Ok(LiteralData::Float {
                significand: integer.into(),
                fraction: fraction_part.into(),
                exponent: InternedStr::default(),
                suffix: InternedStr::default(),
                radix: mode.to_float_radix(),
            });
        };
        
        // advance the state to exponent or suffix
        rest = &fraction_part[non_fraction_start..];
        if rest.starts_with("p") || rest.starts_with("P") {
            debug_assert_eq!(mode, Mode::Hex);
            state = State::Exponent;
        } else if rest.starts_with("e") || rest.starts_with("E") {
            debug_assert_eq!(mode, Mode::Dec);
            state = State::Exponent
        } else {
            if mode == Mode::Hex {
                return Err(NumericLiteralError::MissingHexExponent);
            }

            state = State::Suffix
        }

        &fraction_part[0..non_fraction_start]
    } else {
        ""
    };

    let exponent = if state == State::Exponent {
        let exponent_part = &rest[1..];
        let non_exponent_start = exponent_part.find(|x: char| !x.is_ascii_digit());

        let Some(non_exponent_start) = non_exponent_start else {
            if exponent_part.is_empty() {
                return Err(NumericLiteralError::EmptyExponent);
            } else {
                return Ok(LiteralData::Float {
                    significand: integer.into(),
                    fraction: fraction.into(),
                    exponent: exponent_part.into(),
                    suffix: InternedStr::default(),
                    radix: mode.to_float_radix(),
                });
            }
        };

        rest = &exponent_part[non_exponent_start..];
        state = State::Suffix;

        &exponent_part[0..non_exponent_start]
    } else {
        ""
    };

    let suffix = rest;

    if is_float {
        return Ok(LiteralData::Float {
            significand: integer.into(),
            fraction: fraction.into(),
            exponent: exponent.into(),
            suffix: suffix.into(),
            radix: mode.to_float_radix(),
        });
    } else {
        return Ok(LiteralData::Integer {
            value: integer.into(),
            suffix: suffix.into(),
            radix: mode.to_integer_radix(),
        });
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_integer_literals() {
        assert_eq!(
            number_literal("100"),
            Ok(
                LiteralData::Integer {
                    value: "100".into(),
                    suffix: "".into(),
                    radix: IntegerRadix::Decimal,
                }
            )
        );

        assert_eq!(
            number_literal("0"),
            Ok(
                LiteralData::Integer {
                    value: "0".into(),
                    suffix: "".into(),
                    radix: IntegerRadix::Decimal,
                }
            )
        );

        assert_eq!(
            number_literal("01"),
            Ok(
                LiteralData::Integer {
                    value: "1".into(),
                    suffix: "".into(),
                    radix: IntegerRadix::Octal,
                }
            )
        );

        assert_eq!(
            number_literal("1abcde123"),
            Ok(
                LiteralData::Integer {
                    value: "1".into(),
                    suffix: "abcde123".into(),
                    radix: IntegerRadix::Decimal,
                }
            )
        );

        assert_eq!(
            number_literal("0x1abcde123!"),
            Ok(
                LiteralData::Integer {
                    value: "1abcde123".into(),
                    suffix: "!".into(),
                    radix: IntegerRadix::Hexadecimal,
                }
            )
        );
    }

    #[test]
    fn test_float_literals() {
        assert_eq!(
            number_literal("1.1"),
            Ok(
                LiteralData::Float {
                    significand: "1".into(),
                    fraction: "1".into(),
                    exponent: "".into(),
                    suffix: "".into(),
                    radix: FloatRadix::Decimal,
                }
            )
        );

        assert_eq!(
            number_literal("1.1e5"),
            Ok(
                LiteralData::Float {
                    significand: "1".into(),
                    fraction: "1".into(),
                    exponent: "5".into(),
                    suffix: "".into(),
                    radix: FloatRadix::Decimal,
                }
            )
        );

        assert_eq!(
            number_literal("1.1e5zzz"),
            Ok(
                LiteralData::Float {
                    significand: "1".into(),
                    fraction: "1".into(),
                    exponent: "5".into(),
                    suffix: "zzz".into(),
                    radix: FloatRadix::Decimal,
                }
            )
        );

        assert_eq!(
            number_literal("0x1.1p5"),
            Ok(
                LiteralData::Float {
                    significand: "1".into(),
                    fraction: "1".into(),
                    exponent: "5".into(),
                    suffix: "".into(),
                    radix: FloatRadix::Hexadecimal,
                }
            )
        );
    }

    #[test]
    fn test_float_literals_combo() {
        use std::iter::repeat;
        use itertools::Itertools;

        let iter = repeat([true, false].into_iter()).take(4).multi_cartesian_product();
        for vec in iter {
            let is_hex = vec[0];
            let has_fraction = vec[1];
            let has_exponent = vec[2];
            let has_suffix = vec[3];
            
            let sample = format!(
                "{}{}{}{}{}",
                if is_hex { "0x1" } else { "1" },
                if has_fraction { ".1" } else { "." },
                if has_exponent {
                    if is_hex { "p" } else { "e" }
                } else {
                    ""
                },
                if has_exponent { "5" } else { "" },
                if has_suffix { "asd" } else { "" },
            );

            insta::with_settings!({
                info => &sample, // the template context
                description => "", // the template source code
                omit_expression => true // do not include the default expression
            }, {
                insta::assert_debug_snapshot!(number_literal(&sample));
            });
        }
    }
}