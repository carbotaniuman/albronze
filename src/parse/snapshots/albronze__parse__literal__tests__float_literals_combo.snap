---
source: src/parse/literal.rs
description: ""
info: "0x1.1p5asd"
---
Ok(
    Float {
        significand: "1",
        fraction: "1",
        exponent: "5",
        suffix: "asd",
        radix: Hexadecimal,
    },
)
